### PR TITLE
Review of Open Source Libraries 

### PR DESCRIPTION
**Pull request by hedonistrh**
In this initial version of review of open source projects, I have added Melody RNN and Music VAE. More of them will come. 💪 All feedbacks are welcome.

### CODE REVIEW

**Comment by sertansenturk** [comment]
How many points is this ticket?

**Comment by magdalenafuentes** [comment]
Hey Halil! First thanks for all the work :)
Here's my suggestion: make this shorter and strait to the point. If you want busy people (like Umut) to give valuable feedback on this, you should make it suuuper easy to understand/read, so they can understand the model and comment. 

This is what I mean: https://docs.google.com/presentation/d/1QDWf9p8UVWQaXB3wAPuB4DPZyqWElRQgHSsIsHyJ9lM/edit?usp=sharing

Then use this summary for you to remember more details. It's hard to figure out which model would be useful and why this way. Hope this is useful.

**Comment by hedonistrh** [comment]
Hey Magdalena, thanks for advice. I had no time to simplify my both summary. Also, before the presentation, I would like to reproduce their result with simplest example. So that, I have started with Magenta's RNN.

But, your _presentation idea_ makes sense and I will update it with the summary. _(I wont add any code into presentation.  The markdown file will include development staff which is for you, Sertan and me.)_

**Comment by magdalenafuentes** [comment]
Ok, that sounds great. I'll be checking the slides. Let me know how the experiment goes!

**Comment by hedonistrh** [comment]
Hey @magdalenafuentes, I have add summaries of the libraries into presentation. 

_Ps. Everyone with the link can modify it. May you add limitation?_

**Comment by magdalenafuentes** [comment]
Hey @hedonistrh, yep, I'll take care of this on Monday. Also I'll check the evaluation metric summary that day. Thanks!

**Comment by magdalenafuentes** [comment]
Hey @hedonistrh, just went over your slides, they look great! I corrected a couple of details and added comments on red about monophonic/polyphonic stuff we should keep in mind in the discussion (small detail, the slides seem repeated twice, do you see the same?).

About the limitations, I'm not clear what you had in mind, but it seems to me that we should pick one/two models and start working with them, then read more in detail and point out the limitations from the experiments? I don't really know what would come out from these models when applied to Turkish music. What do you think @sertansenturk ? Let's move to the discussion stage?

**Comment by hedonistrh** [comment]
Hey @magdalenafuentes , thank you so much 🌱 I will check especially _MusicVAE_ for monophonic generation. 

Firstly, I think, we should start with Magenta RNN which has 3 different model and pre-trained models. _(At this night, I will start to work with it. Hopefully, until next meeting, we will have some conclusion about it)_ @sertansenturk 's comment can be so helpful. 👍

**Comment by magdalenafuentes** [comment]
Good! BTW, I saw you comment you were having troubles with the conda environment that Magenta provides. It would be great if it works (because it's easy to make it work in the cluster here), so if you want, share the errors you got to see if we can help.

**Comment by hedonistrh** [comment]
Hey @magdalenafuentes, when I tried _Conda_ version of _Magenta_, it installed _miniconda2_ and some packages. After that, I have tried to generate sample, however, I got error about missing libraries. I installed that one, again I got that error. It repeats 4 times, so, I just switched to _Docker_.

I have not worked it before. So, I do not know that how we use Docker with cluster. 🙍 

According to my first research, we can use [Docker Swarm](https://docs.docker.com/v17.12/get-started/part4/#set-up-your-swarm). I will check it. Also, maybe @sertansenturk can help 🌻

**Comment by hedonistrh** [comment]
I uploaded _.md_ file about my first experiments with magenta RNN. There is bunch of problem with our input. In the next meeting, we can talk about that. 

From _musicXML_ perspective, it give a warning for each file.

> WARNING:tensorflow:Could not parse MusicXML file /magenta-data/MusicXML/hicaz--sarki--aksak--kis_geldi--sevki_bey.xml. It will be skipped. Error was: Could not find fifths attribute in key signature.

From midi perspective, why it discard so many inputs because of _polyphony_? 😕  Are there any _polyphony_ music in our dataset? @sertansenturk @barisbozkurt 

screen shot 2019-02-19 at 22 18 12

**Comment by sertansenturk** [comment]
> Hey @magdalenafuentes, when I tried _Conda_ version of _Magenta_, it installed _miniconda2_ and some packages. After that, I have tried to generate sample, however, I got error about missing libraries. I installed that one, again I got that error. It repeats 4 times, so, I just switched to _Docker_.
> 
> I have not worked it before. So, I do not know that how we use Docker with cluster. 
> 
> According to my first research, we can use [Docker Swarm](https://docs.docker.com/v17.12/get-started/part4/#set-up-your-swarm). I will check it. Also, maybe @sertansenturk can help

Docker is the best option if the cluster supports and @magdalenafuentes has experience with it. Otherwise, conda is the most feasible option. Docker Swarm is mostly irrelevant and impractical for our purposes (see below for your information).

==== 

Docker Swarm (or better, Kubernetes) would be used typically when there are multiple Docker containers, which require each others output. In such a case you would need a container orchestrator, which monitors each container's activity and coordinate/manage the operation. I think we neither have a need for such complexity in our experimental setup, nor we would be able to set it up due to lack of time, resource & cluster flexibility.

**Comment by sertansenturk** [comment]
> I uploaded _.md_ file about my first experiments with magenta RNN. There is a bunch of problem with our input. In the next meeting, we can talk about that.
> 
> From _musicXML_ perspective, it give a warning for each file.
> 
> > WARNING:tensorflow:Could not parse MusicXML file /magenta-data/MusicXML/hicaz--sarki--aksak--kis_geldi--sevki_bey.xml. It will be skipped. Error was: Could not find fifths attribute in key signature.
> 
> From midi perspective, why it discard so many inputs because of _polyphony_?  Are there any _polyphony_ music in our dataset? @sertansenturk @barisbozkurt
> 
> screen shot 2019-02-19 at 22 18 12

On the MusicXML side, magenta RNN seem to require the Western classical music key signatures. Makam music have additional key signatures. We may try to replace the key signatures with blank (C major) key signature (we have to figure out how the model uses the key signatures), and see if the model may then parse the scores. @hedonistrh do not try to do it for all the scores, just pick a few, replace the key signature, and then try if modified scores are parsed properly (I doubt it'll be the case though).

The SymbTr-MIDI scores are monophonic, therefore the preprocessing code most likely raises an incorrect polyphony error. I suspect two potential reasons: 1) I don't think if the instrument information is encoded in the MIDI files, the preprocessing code might be going crazy because of  its lack, 2) the note tuning is stored as pitch bends, and the code may not be able to handle it.

**Comment by sertansenturk** [comment]
I have also read the presentation, good job.

It looks we will spend some time on reading data. @hedonistrh how close are you to tackle the I/O question? If you believe it will take some time, we merge this pull request, close the ticket, and open separate tickets for each model.

**Comment by hedonistrh** [comment]
I think, we have to tackle with why _Magenta_ throw an exception for our dataset as _polyphony._ For this PR, I will discard the markdown file which is related with experiments. Lets merge this PR with just review of open source libraries. With time, maybe, we can expand it.

**Comment by sertansenturk** [comment]
> I think, we have to tackle with why _Magenta_ throw an exception for our dataset as _polyphony._ For this PR, I will discard the markdown file which is related with experiments. Lets merge this PR with just review of open source libraries. With time, maybe, we can expand it.

I would argue on not including the experiment-related experiments: you have already done the preliminary work, so why do we remove that!? __9 trials would lead the 10th to be correct solution,__ so we should include your trial on the path.

Instead, I would ask from you to explicitly state that __the trial has failed, and why in the start of the document__...

**Comment by hedonistrh** [comment]
- I just remove it for this PR. I suggest that we can discuss experiment results in the different PR. But if you want to keep open this PR, I can add it. :)

- I was writing the .md file which is about experiments in chronological order and I just started with **conda** (which did not work). Thus, it is in the **start** of the document.

**Comment by hedonistrh** [comment]
I think that we can merge this PR at this moment. If we need to add more libraries, we can change it later. 👍

**Comment by magdalenafuentes** [comment]
Agree!

**Comment by hedonistrh** [comment]
@sertansenturk Should we merge this? 🎐
